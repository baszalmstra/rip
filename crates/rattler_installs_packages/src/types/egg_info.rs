use crate::types::{NormalizedPackageName, ParsePackageNameError};
use pep440_rs::Version;
use std::fmt::{Display, Formatter};
use std::str::FromStr;
use thiserror::Error;

/// The type of extension for an egg.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum EggExt {
    /// `.egg` extension
    Egg,

    /// `.egg-info` extension
    EggInfo,
}

impl EggExt {
    /// Returns the extension as a string (without the preceding `.` ).
    pub fn as_str(&self) -> &str {
        match self {
            EggExt::Egg => "egg-info",
            EggExt::EggInfo => "egg",
        }
    }
}

/// Represents an egg filename.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EggFilename {
    /// The name of the package
    pub name: NormalizedPackageName,

    /// Optional version of the package
    pub version: Option<Version>,

    /// Optional required python version
    pub python_version: Option<String>,

    /// Optional required platform
    pub required_platform: Option<String>,

    /// The extension
    pub extension: EggExt,
}

impl Display for EggFilename {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name.as_str())?;
        if let Some(version) = &self.version {
            write!(f, "-{version}")?;
        }
        if let Some(python_version) = &self.python_version {
            write!(f, "-py{python_version}")?;
        }
        if let Some(required_platform) = &self.required_platform {
            write!(f, "-{required_platform}")?;
        }
        write!(f, ".{}", self.extension.as_str())
    }
}

/// An error that might occur when parsing an [`EggFilename`].
#[derive(Debug, Error)]
pub enum InvalidEggFilename {
    /// The extension of the filename is not recognized
    #[error("the extension is unrecognized, only '.egg' and '.egg-info' supported")]
    UnrecognizedExtension,

    /// Could not find a required package name in the filename
    #[error("the filename is missing a package name")]
    MissingName,

    /// Could not parse the package name as a valid normalized package name.
    #[error("{0}")]
    InvalidPackageName(ParsePackageNameError),

    /// The version is invalid
    #[error("invalid version: {0}")]
    InvalidVersion(String),

    /// The python version in the filename is invalid.
    #[error("invalid python version, must start with 'py'")]
    InvalidPythonVersion,
}

impl FromStr for EggFilename {
    type Err = InvalidEggFilename;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // Parse the extension
        let (extension, rest) = if let Some(rest) = s.strip_suffix(".egg") {
            (EggExt::Egg, rest)
        } else if let Some(rest) = s.strip_suffix(".egg-info") {
            (EggExt::EggInfo, rest)
        } else {
            return Err(InvalidEggFilename::UnrecognizedExtension);
        };

        // Split the rest of the string based on dashes.
        let mut split = rest.split('-');

        // Parse the name of the package
        let name = split
            .next()
            .ok_or(InvalidEggFilename::MissingName)
            .and_then(|s| {
                NormalizedPackageName::from_str(s).map_err(InvalidEggFilename::InvalidPackageName)
            })?;

        // Parse the optional version
        let version = split
            .next()
            .map(Version::from_str)
            .transpose()
            .map_err(InvalidEggFilename::InvalidVersion)?;

        // Parse the optional python version
        let python_version = match split.next().map(|s| s.strip_prefix("py")) {
            Some(Some(py_version)) => Some(py_version.to_string()),
            Some(None) => return Err(InvalidEggFilename::InvalidPythonVersion),
            None => None,
        };

        // Parse the required platform
        let required_platform = split.next().map(ToString::to_string);

        Ok(Self {
            name,
            version,
            python_version,
            required_platform,
            extension,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use indexmap::IndexMap;

    #[test]
    fn test_egg_file_name() {
        let cases = [
            "simplejson-2.2.1-py3.9.egg-info",
            "pywin32-213-py2.6.egg",
            "simple.egg-info",
            "hunspell-0.3.3.egg-info",
            "pygobject-3.14.0-py2.7-linux-x86_64.egg-info",
        ];

        let conversion = cases
            .into_iter()
            .map(|s| (s, EggFilename::from_str(s)))
            .collect::<IndexMap<_, _>>();
        insta::assert_debug_snapshot!(conversion);
    }
}
