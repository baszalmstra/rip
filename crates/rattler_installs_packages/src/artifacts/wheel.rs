use crate::types::HasArtifactName;
use crate::{
    types::{
        ArtifactFromBytes, NormalizedPackageName, PackageName, RFC822ish, WheelCoreMetaDataError,
        WheelCoreMetadata, WheelFilename,
    },
    utils::ReadAndSeek,
};
use async_http_range_reader::AsyncHttpRangeReader;
use async_zip::base::read::seek::ZipFileReader;
use fs_err as fs;
use miette::IntoDiagnostic;
use parking_lot::Mutex;
use pep440_rs::Version;
use std::{
    borrow::Cow,
    ffi::OsStr,
    io::Read,
    path::{Path, PathBuf},
    str::FromStr,
};
use thiserror::Error;
use tokio_util::compat::TokioAsyncReadCompatExt;
use zip::{result::ZipError, ZipArchive};

/// Different representations of a wheel that is stored locally on disk.
pub enum LocalWheel {
    /// A `.whl` file on disk.
    ArchivedWheel(ArchivedWheel),
}

/// A wheel file (`.whl`) in its archived form that is stored somewhere on disk.
pub struct ArchivedWheel {
    /// Name of wheel
    pub name: WheelFilename,

    pub(crate) archive: Mutex<ZipArchive<Box<dyn ReadAndSeek + Send>>>,
}

impl HasArtifactName for ArchivedWheel {
    type Name = WheelFilename;

    fn name(&self) -> &Self::Name {
        &self.name
    }
}

impl HasArtifactName for LocalWheel {
    type Name = WheelFilename;

    fn name(&self) -> &Self::Name {
        match self {
            LocalWheel::ArchivedWheel(whl) => whl.name(),
        }
    }
}

impl ArtifactFromBytes for ArchivedWheel {
    fn from_bytes(name: Self::Name, bytes: Box<dyn ReadAndSeek + Send>) -> miette::Result<Self> {
        Ok(Self {
            name,
            archive: Mutex::new(ZipArchive::new(bytes).into_diagnostic()?),
        })
    }
}

impl ArtifactFromBytes for LocalWheel {
    fn from_bytes(name: Self::Name, bytes: Box<dyn ReadAndSeek + Send>) -> miette::Result<Self> {
        // It is assumed that when constructing a LocalWheel from bytes that the bytes refer to a
        // zipped archive.
        ArchivedWheel::from_bytes(name, bytes).map(LocalWheel::ArchivedWheel)
    }
}

impl ArchivedWheel {
    /// Open a wheel by reading a `.whl` file on disk.
    pub fn from_path(
        path: &Path,
        normalized_package_name: &NormalizedPackageName,
    ) -> miette::Result<Self> {
        let file_name = path
            .file_name()
            .and_then(OsStr::to_str)
            .ok_or_else(|| miette::miette!("path does not contain a filename"))?;
        let wheel_name =
            WheelFilename::from_filename(file_name, normalized_package_name).into_diagnostic()?;
        let file = fs::File::open(path).into_diagnostic()?;
        Self::from_bytes(wheel_name, Box::new(file))
    }

    /// Create a wheel from URL and content.
    pub fn from_url_and_bytes(
        url: &str,
        normalized_package_name: &NormalizedPackageName,
        bytes: Box<dyn ReadAndSeek + Send>,
    ) -> miette::Result<Self> {
        let url_path = PathBuf::from_str(url).into_diagnostic()?;
        let file_name = url_path
            .file_name()
            .and_then(OsStr::to_str)
            .ok_or_else(|| {
                miette::miette!("path {:?} does not contain a wheel filename", url_path)
            })?;
        let wheel_filename =
            WheelFilename::from_filename(file_name, normalized_package_name).into_diagnostic()?;

        Self::from_bytes(wheel_filename.clone(), Box::new(bytes))
    }

    async fn get_lazy_vitals(
        name: &WheelFilename,
        stream: &mut AsyncHttpRangeReader,
    ) -> Result<(Vec<u8>, WheelCoreMetadata), WheelVitalsError> {
        // Make sure we have the back part of the stream.
        // Best guess for the central directory size inside the zip
        const CENTRAL_DIRECTORY_SIZE: u64 = 16384;
        // Because the zip index is at the back
        stream
            .prefetch(stream.len().saturating_sub(CENTRAL_DIRECTORY_SIZE)..stream.len())
            .await;

        // Construct a zip reader to uses the stream.
        let mut reader = ZipFileReader::new(stream.compat())
            .await
            .map_err(|err| WheelVitalsError::from_async_zip("/".into(), err))?;

        // Collect all top-level filenames
        let file_names = reader
            .file()
            .entries()
            .iter()
            .filter_map(|e| e.filename().as_str().ok());

        // Determine the name of the dist-info directory
        let dist_info_prefix = find_dist_info(&name, file_names)?.to_owned();

        let metadata_path = format!("{dist_info_prefix}.dist-info/METADATA");
        let (metadata_idx, metadata_entry) = reader
            .file()
            .entries()
            .iter()
            .enumerate()
            .find(|(_, p)| p.filename().as_str().ok() == Some(metadata_path.as_str()))
            .ok_or(WheelVitalsError::MetadataMissing)?;

        // Get the size of the entry plus the header + size of the filename. We should also actually
        // include bytes for the extra fields but we don't have that information.
        let offset = metadata_entry.header_offset();
        let size = metadata_entry.compressed_size()
            + 30 // Header size in bytes
            + metadata_entry.filename().as_bytes().len() as u64;

        // The zip archive uses as BufReader which reads in chunks of 8192. To ensure we prefetch
        // enough data we round the size up to the nearest multiple of the buffer size.
        let buffer_size = 8192;
        let size = ((size + buffer_size - 1) / buffer_size) * buffer_size;

        // Fetch the bytes from the zip archive that contain the requested file.
        reader
            .inner_mut()
            .get_mut()
            .prefetch(offset..offset + size)
            .await;

        // Read the contents of the metadata.json file
        let mut contents = Vec::new();
        reader
            .reader_with_entry(metadata_idx)
            .await
            .map_err(|e| WheelVitalsError::from_async_zip(metadata_path.clone(), e))?
            .read_to_end_checked(&mut contents)
            .await
            .map_err(|e| WheelVitalsError::from_async_zip(metadata_path, e))?;

        // Parse the wheel data
        let metadata = WheelCoreMetadata::try_from(contents.as_slice())?;

        let stream = reader.into_inner().into_inner();
        let ranges = stream.requested_ranges().await;
        let total_bytes_fetched: u64 = ranges.iter().map(|r| r.end - r.start).sum();
        tracing::debug!(
            "fetched {} ranges, total of {} bytes, total file length {} ({}%)",
            ranges.len(),
            total_bytes_fetched,
            stream.len(),
            (total_bytes_fetched as f64 / stream.len() as f64 * 100000.0).round() / 100.0
        );

        Ok((contents, metadata))
    }

    /// Get the metadata from the wheel archive
    pub fn metadata(&self) -> Result<(Vec<u8>, WheelCoreMetadata), WheelVitalsError> {
        let mut archive = self.archive.lock();

        // Determine the name of the dist-info directory
        let dist_info_prefix = find_dist_info(&self.name, archive.file_names())?.to_owned();

        // Read the METADATA file
        let metadata_path = format!("{dist_info_prefix}.dist-info/METADATA");
        let metadata_blob = read_entry_to_end(&mut archive, &metadata_path)?;
        let metadata = WheelCoreMetadata::try_from(metadata_blob.as_slice())?;

        // Verify the contents of the METADATA
        if metadata.name != self.name.distribution {
            return Err(WheelCoreMetaDataError::FailedToParse(format!(
                "name mismatch between {dist_info_prefix}.dist-info/METADATA and filename ({} != {}",
                metadata.name.as_source_str(),
                self.name.distribution.as_source_str()
            ))
            .into());
        }
        if metadata.version != self.name.version {
            return Err(WheelCoreMetaDataError::FailedToParse(format!(
                "version mismatch between {dist_info_prefix}.dist-info/METADATA and filename ({} != {})",
                metadata.version, self.name.version
            ))
            .into());
        }

        Ok((metadata_blob, metadata))
    }

    /// Read metadata from bytes-stream
    pub async fn read_metadata_bytes(
        name: &WheelFilename,
        stream: &mut AsyncHttpRangeReader,
    ) -> miette::Result<(Vec<u8>, WheelCoreMetadata)> {
        Self::get_lazy_vitals(name, stream).await.into_diagnostic()
    }
}

#[derive(Debug, Error)]
#[allow(missing_docs)]
pub enum WheelVitalsError {
    #[error(".dist-info/ missing")]
    DistInfoMissing,

    #[error(".dist-info/WHEEL missing")]
    WheelMissing,

    #[error(".dist-info/METADATA missing")]
    MetadataMissing,

    #[error("found multiple {0} directories in wheel")]
    MultipleSpecialDirs(Cow<'static, str>),

    #[error("failed to parse WHEEL file")]
    FailedToParseWheel(#[source] <RFC822ish as FromStr>::Err),

    #[error("unsupported WHEEL version {0}")]
    UnsupportedWheelVersion(String),

    #[error("invalid METADATA")]
    InvalidMetadata(#[from] WheelCoreMetaDataError),

    #[error(transparent)]
    IoError(#[from] std::io::Error),

    #[error("Failed to read the wheel file {0}")]
    ZipError(String, #[source] ZipError),

    #[error("Failed to read the wheel file {0}: {1}")]
    AsyncZipError(String, #[source] async_zip::error::ZipError),

    #[error("missing key from WHEEL '{0}'")]
    MissingKeyInWheel(String),
}

impl WheelVitalsError {
    pub(crate) fn from_zip(file: String, err: ZipError) -> Self {
        match err {
            ZipError::Io(err) => WheelVitalsError::IoError(err),
            ZipError::FileNotFound => {
                if file.ends_with("WHEEL") {
                    WheelVitalsError::WheelMissing
                } else if file.ends_with("METADATA") {
                    WheelVitalsError::MetadataMissing
                } else {
                    WheelVitalsError::ZipError(file, err)
                }
            }
            _ => WheelVitalsError::ZipError(file, err),
        }
    }

    pub(crate) fn from_async_zip(file: String, err: async_zip::error::ZipError) -> Self {
        match err {
            async_zip::error::ZipError::UpstreamReadError(err) => WheelVitalsError::IoError(err),
            _ => WheelVitalsError::AsyncZipError(file, err),
        }
    }
}

/// Helper method to read a particular file from a zip archive.
fn read_entry_to_end<R: ReadAndSeek>(
    archive: &mut ZipArchive<R>,
    name: &str,
) -> Result<Vec<u8>, WheelVitalsError> {
    let mut bytes = Vec::new();
    archive
        .by_name(name)
        .map_err(|err| WheelVitalsError::from_zip(name.to_string(), err))?
        .read_to_end(&mut bytes)?;

    Ok(bytes)
}

/// Locates the `.dist-info` directory in a list of files.
pub(crate) fn find_dist_info<'a>(
    wheel_name: &WheelFilename,
    files: impl IntoIterator<Item = &'a str>,
) -> Result<&'a str, WheelVitalsError> {
    let mut dist_infos = files.into_iter().filter_map(|path| {
        let (dir_name, rest) = path.split_once(['/', '\\'])?;
        let dir_stem = dir_name.strip_suffix(".dist-info")?;
        let (name, version) = dir_stem.rsplit_once('-')?;
        if PackageName::from_str(name).ok()? == wheel_name.distribution
            && Version::from_str(version).ok()? == wheel_name.version
            && rest == "METADATA"
        {
            Some(dir_stem)
        } else {
            None
        }
    });

    match (dist_infos.next(), dist_infos.next()) {
        (Some(path), None) => Ok(path),
        (Some(_), Some(_)) => Err(WheelVitalsError::MultipleSpecialDirs("dist-info".into())),
        _ => Err(WheelVitalsError::DistInfoMissing),
    }
}
