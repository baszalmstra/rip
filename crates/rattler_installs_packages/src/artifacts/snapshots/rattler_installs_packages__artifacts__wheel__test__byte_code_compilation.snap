---
source: crates/rattler_installs_packages/src/artifacts/wheel.rs
assertion_line: 1346
expression: record_content
---
../../../bin/cowpy,sha256=SFF6UwGGzU1ml9elXqzPZDc32EjsJpLKC8MoU7S4yHI,200
cowpy-1.1.5.dist-info/INSTALLER,sha256=FpzDrzfcP6oI1mbsvYAjw7hT2I4P4wr99k5NKqmKJt0,10
cowpy-1.1.5.dist-info/LICENSE.md,sha256=aP4kRCPk1Pc11h2VaqptJ3jejrF5N0JPNQvM34wCFoo,11346
cowpy-1.1.5.dist-info/METADATA,sha256=2hFQ8fNsIneBFh9xNZGgTZyuhY9op1GgBA5cWCi7iNY,438
cowpy-1.1.5.dist-info/RECORD,,
cowpy-1.1.5.dist-info/WHEEL,sha256=y3eDiaFVSNTPbgzfNn0nYn5tEn1cX6WrdetDlQM4xWw,83
cowpy-1.1.5.dist-info/entry_points.txt,sha256=65rFxa9IJJm7mT1X4sKaKZ3AhxSPukmKDXlXwOnVxOM,40
cowpy/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
cowpy/__pycache__/__init__.cpython-<version>.pyc,,
cowpy/__pycache__/cow.cpython-<version>.pyc,,
cowpy/cow.py,sha256=W9eX898X1ocpWGvF1SrBOEEiY7z8tfglxFmtmGrRwxw,41928
